README.md
# rsql-ruby

Ruby RSQL/FIQL parser library. 

Kudos:

* [Java RSQL parser](https://github.com/jirutka/rsql-parser)
* [Node FIQL parser](https://github.com/vmoebius/node-fiql-parser)

## How to use?

```
gem install rsql-ruby
```

```
require 'rsql-ruby'
puts RsqlRuby.parse("movies=in=(Matrix, Scary Movie),year=gte=2020")
```

for more see tests.

## Contributions

Just open pull request with changes and proper test.

# LICENSE:

(The MIT License)

Copyright (c) 2020 Strnadj

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


History.txt
=== 1.0.0 / 2020-05-24

* First version



test/rsql_ruby/test_fiql.rb
gem 'minitest'
require 'minitest/autorun'
require 'rsql_ruby'

module TestRsqlRuby; end

class TestRsqlRuby::TestFiql < Minitest::Test
  def test_basic_expressions
    parser = RsqlRuby::Parser.new

    {
      "a=eq=b": { type: :CONSTRAINT, selector: 'a', comparsion: '=eq=', argument: 'b' },
      "a=custom=b": { type: :CONSTRAINT, selector: 'a', comparsion: '=custom=', argument: 'b' },
      "a==b": { type: :CONSTRAINT, selector: 'a', comparsion: '==', argument: 'b' },
      "a==2018-09-01T12:14:28Z": { type: :CONSTRAINT, selector: 'a', comparsion: '==', argument: '2018-09-01T12:14:28Z' },
      "a!=b": { type: :CONSTRAINT, selector: 'a', comparsion: '!=', argument: 'b' },
      "field=op=(item0,item1,item2)": { type: :CONSTRAINT, selector: 'field', comparsion: '=op=', argument: ['item0', 'item1', 'item2'] },
    }.each do |key, output|
      assert_equal(parser.parse(key.to_s), output)
    end
  end

  def test_combinations
    {
      "a=eq=b;c=ne=d" => {
        type: :COMBINATION,
        operator: :AND,
        lhs: { type: :CONSTRAINT, selector: 'a', comparsion: '=eq=', argument: 'b' },
        rhs: { type: :CONSTRAINT, selector: 'c', comparsion: '=ne=', argument: 'd' },
      },
      "a=eq=b;c=ne=d;e=gt=f" => {
        type: :COMBINATION,
        operator: :AND,
        lhs: { type: :CONSTRAINT, selector: 'a', comparsion: '=eq=', argument: 'b' },
        rhs: {
          type: :COMBINATION,
          operator: :AND,
          lhs: { type: :CONSTRAINT, selector: 'c', comparsion: '=ne=', argument: 'd' },
          rhs: { type: :CONSTRAINT, selector: 'e', comparsion: '=gt=', argument: 'f' }
        }
      },
      "a=eq=b,c=ne=d" => {
        type: :COMBINATION,
        operator: :OR,
        lhs: { type: :CONSTRAINT, selector: 'a', comparsion: '=eq=', argument: 'b' },
        rhs: { type: :CONSTRAINT, selector: 'c', comparsion: '=ne=', argument: 'd' },
      },
      "a=eq=b,c=ne=d,e=gt=(f,g,c)" => {
        type: :COMBINATION,
        operator: :OR,
        lhs: { type: :CONSTRAINT, selector: 'a', comparsion: '=eq=', argument: 'b' },
        rhs: {
          type: :COMBINATION,
          operator: :OR,
          lhs: { type: :CONSTRAINT, selector: 'c', comparsion: '=ne=', argument: 'd' },
          rhs: { type: :CONSTRAINT, selector: 'e', comparsion: '=gt=', argument: ['f', 'g', 'c'] }
        }
      },
      "a=eq=b;c=ne=d,e=gt=f" => { # OR BEFORE AND
        type: :COMBINATION,
        operator: :OR,
        lhs: {
          type: :COMBINATION,
          operator: :AND,
          lhs: { type: :CONSTRAINT, selector: 'a', comparsion: '=eq=', argument: 'b' },
          rhs: { type: :CONSTRAINT, selector: 'c', comparsion: '=ne=', argument: 'd' }
        },
        rhs: { type: :CONSTRAINT, selector: 'e', comparsion: '=gt=', argument: 'f' },
      },
      'a=eq=b;(c=ne=d,e=gt=f)' => { # GROUPING
        type: :COMBINATION,
        operator: :AND,
        lhs: { type: :CONSTRAINT, selector: 'a', comparsion: '=eq=', argument: 'b' },
        rhs: {
          type: :COMBINATION,
          operator: :OR,
          lhs: { type: :CONSTRAINT, selector: 'c', comparsion: '=ne=', argument: 'd' },
          rhs: { type: :CONSTRAINT, selector: 'e', comparsion: '=gt=', argument: 'f' }
        }
      }
    }.each do |key, output|
      assert_equal(RsqlRuby.parse(key.to_s), output)
    end
  end

end

test/rsql_ruby/test_rsql.rb
gem 'minitest'
require 'minitest/autorun'
require 'rsql_ruby'

module TestRsqlRuby; end

class TestRsqlRuby::TestRsql < Minitest::Test
  def test_basic_expressions
    {
      'name=="Kill Bill";year=gt=2003' => {
        type: :COMBINATION,
        operator: :AND,
        lhs: { type: :CONSTRAINT, selector: 'name', comparsion: '==', argument: 'Kill Bill' },
        rhs: { type: :CONSTRAINT, selector: 'year', comparsion: '=gt=', argument: '2003' },
      },
      'name=="Kill;nam \'\'e=gt=Bill";year=gt=2003' => {
        type: :COMBINATION,
        operator: :AND,
        lhs: { type: :CONSTRAINT, selector: 'name', comparsion: '==', argument: 'Kill;nam \'\'e=gt=Bill' },
        rhs: { type: :CONSTRAINT, selector: 'year', comparsion: '=gt=', argument: '2003' },
      },
      'name==\'Kill;"name"=gt=Bill\',year=gt=2003' => {
        type: :COMBINATION,
        operator: :OR,
        lhs: { type: :CONSTRAINT, selector: 'name', comparsion: '==', argument: 'Kill;"name"=gt=Bill' },
        rhs: { type: :CONSTRAINT, selector: 'year', comparsion: '=gt=', argument: '2003' },
      },
      "name=in=('sci-fi','action');year=gt=2003" => {
        type: :COMBINATION,
        operator: :AND,
        lhs: { type: :CONSTRAINT, selector: 'name', comparsion: '=in=', argument: [ 'sci-fi', 'action' ] },
        rhs: { type: :CONSTRAINT, selector: 'year', comparsion: '=gt=', argument: '2003' },
      },
      'name=in="test \' \" name"' => {
        type: :CONSTRAINT,
        selector: 'name' ,
        comparsion: '=in=',
        argument: 'test \' \\" name',
      }
    }.each do |key, output|
      assert_equal(RsqlRuby.parse(key.to_s), output)
    end
  end
end

lib/rsql_ruby/lexer.rex.rb
# encoding: UTF-8
#--
# This file is automatically generated. Do not modify it.
# Generated by: oedipus_lex version 2.5.1.
# Source: lib/rsql_ruby/lexer.rex
#++

module RsqlRuby


##
# The generated lexer Parser < Racc::Parser

class Parser < Racc::Parser
  require 'strscan'

  # :stopdoc:
  BLANK                = /(\ |\t)+/
  UNRESERVED           = /[a-zA-Z0-9\-._~:]+/
  COMPARATOR           = /!=|(=([a-z]|!)*=)/
  AND_OPERATOR         = /;/
  OR_COMMA_OPERATOR    = /,/
  SINGLE_QUOTE         = /'/
  DOUBLE_QUOTE         = /"/
  SINGLE_QUOTED_STRING = /[^']+/
  DOUBLE_QUOTED_STRING = /[^"]+/
  OPENING_BRACKET      = /\(/
  CLOSING_BRACKET      = /\)/
  # :startdoc:
  # :stopdoc:
  class LexerError < StandardError ; end
  class ScanError < LexerError ; end
  # :startdoc:

  ##
  # The file name / path

  attr_accessor :filename

  ##
  # The StringScanner for this lexer.

  attr_accessor :ss

  ##
  # The current lexical state.

  attr_accessor :state

  alias :match :ss

  ##
  # The match groups for the current scan.

  def matches
    m = (1..9).map { |i| ss[i] }
    m.pop until m[-1] or m.empty?
    m
  end

  ##
  # Yields on the current action.

  def action
    yield
  end


  ##
  # The current scanner class. Must be overridden in subclasses.

  def scanner_class
    StringScanner
  end unless instance_methods(false).map(&:to_s).include?("scanner_class")

  ##
  # Parse the given string.

  def parse str
    self.ss     = scanner_class.new str
    self.state  ||= nil

    do_parse
  end

  ##
  # Read in and parse the file at +path+.

  def parse_file path
    self.filename = path
    open path do |f|
      parse f.read
    end
  end

  ##
  # The current location in the parse.

  def location
    [
      (filename || "<input>"),
    ].compact.join(":")
  end

  ##
  # Lex the next token.

  def next_token

    token = nil

    until ss.eos? or token do
      token =
        case state
        when nil then
          case
          when ss.skip(/#{BLANK}/) then
            # do nothing
          when text = ss.scan(/#{UNRESERVED}/) then
            action { [:UNRESERVED, text] }
          when text = ss.scan(/#{COMPARATOR}/) then
            action { [:COMPARATOR, text] }
          when text = ss.scan(/#{OR_COMMA_OPERATOR}/) then
            action { [:OR_COMMA_OPERATOR, text] }
          when text = ss.scan(/#{AND_OPERATOR}/) then
            action { [:AND_OPERATOR, text] }
          when text = ss.scan(/#{OPENING_BRACKET}/) then
            action { [:OPENING_BRACKET, text] }
          when text = ss.scan(/#{CLOSING_BRACKET}/) then
            action { [:CLOSING_BRACKET, text] }
          when text = ss.scan(/#{SINGLE_QUOTE}/) then
            action { self.state = :SINGLE_QUOTE; [:SINGLE_QUOTE, text] }
          when text = ss.scan(/#{DOUBLE_QUOTE}/) then
            action { self.state = :DOUBLE_QUOTE; [:DOUBLE_QUOTE, text] }
          else
            text = ss.string[ss.pos .. -1]
            raise ScanError, "can not match (#{state.inspect}) at #{location}: '#{text}'"
          end
        when :SINGLE_QUOTE then
          case
          when text = ss.scan(/#{SINGLE_QUOTED_STRING}/) then
            action { [:SINGLE_QUOTED_STRING, text] }
          when text = ss.scan(/#{SINGLE_QUOTE}/) then
            action { self.state = nil; [:SINGLE_QUOTE, text] }
          else
            text = ss.string[ss.pos .. -1]
            raise ScanError, "can not match (#{state.inspect}) at #{location}: '#{text}'"
          end
        when :DOUBLE_QUOTE then
          case
          when text = ss.scan(/#{DOUBLE_QUOTED_STRING}/) then
            action { [:DOUBLE_QUOTED_STRING, text] }
          when text = ss.scan(/#{DOUBLE_QUOTE}/) then
            action { self.state = nil; [:DOUBLE_QUOTE, text] }
          else
            text = ss.string[ss.pos .. -1]
            raise ScanError, "can not match (#{state.inspect}) at #{location}: '#{text}'"
          end
        else
          raise ScanError, "undefined state at #{location}: '#{state}'"
        end # token = case state

      next unless token # allow functions to trigger redo w/ nil
    end # while

    raise LexerError, "bad lexical result at #{location}: #{token.inspect}" unless
      token.nil? || (Array === token && token.size >= 2)

    # auto-switch state
    self.state = token.last if token && token.first == :state

    token
  end # def next_token
end # class

  end


lib/rsql_ruby/parser.rb
#
# DO NOT MODIFY!!!!
# This file is automatically generated by Racc 1.5.0
# from Racc grammar file "".
#

require 'racc/parser.rb'

  # Generated by racc
  require_relative 'lexer.rex'

module RsqlRuby
  class Parser < Racc::Parser


def create_logical_operator(operator, lhs, rhs)
  {
    type: :COMBINATION,
    operator: operator,
    lhs: lhs,
    rhs: rhs
  }
end

def create_constraint(selector, comparsion, argument)
  {
    type: :CONSTRAINT,
    selector: selector,
    comparsion: comparsion,
    argument: argument
  }
end

def create_array(string)

end

##### State transition tables begin ###

racc_action_table = [
    19,    11,    20,    21,    30,    22,    27,    28,    29,    25,
    20,    21,     8,    22,    20,    21,     6,    22,     7,     6,
     6,     7,     7,     6,    31,     7,    32,     9,    13,    10,
    23 ]

racc_action_check = [
    11,     5,    11,    11,    24,    11,    21,    22,    24,    19,
    19,    19,     1,    19,    30,    30,    10,    30,    10,     6,
     9,     6,     9,     0,    27,     0,    28,     3,     8,     4,
    12 ]

racc_action_pointer = [
    16,    12,   nil,    23,    24,    -5,    12,   nil,    28,    13,
     9,    -7,    22,   nil,   nil,   nil,   nil,   nil,   nil,     1,
   nil,    -5,    -6,   nil,     0,   nil,   nil,    14,    14,   nil,
     5,   nil,   nil,   nil ]

racc_action_default = [
   -19,   -19,    -1,    -3,    -5,    -7,   -19,    -9,   -19,   -19,
   -19,   -19,   -19,    34,    -2,    -4,    -6,   -10,   -11,   -19,
   -16,   -19,   -19,    -8,   -19,   -13,   -14,   -19,   -19,   -12,
   -19,   -17,   -18,   -15 ]

racc_goto_table = [
    18,     2,     1,    15,    16,    17,    24,    12,    26,   nil,
    14,   nil,   nil,   nil,   nil,   nil,   nil,   nil,   nil,    33 ]

racc_goto_check = [
     8,     2,     1,     3,     6,     7,     9,     2,     8,   nil,
     2,   nil,   nil,   nil,   nil,   nil,   nil,   nil,   nil,     8 ]

racc_goto_pointer = [
   nil,     2,     1,    -7,   nil,   nil,    -7,    -6,   -11,   -13 ]

racc_goto_default = [
   nil,   nil,   nil,     3,     4,     5,   nil,   nil,   nil,   nil ]

racc_reduce_table = [
  0, 0, :racc_error,
  1, 15, :_reduce_none,
  3, 16, :_reduce_2,
  1, 16, :_reduce_none,
  3, 17, :_reduce_4,
  1, 17, :_reduce_none,
  3, 18, :_reduce_6,
  1, 18, :_reduce_none,
  3, 18, :_reduce_8,
  1, 19, :_reduce_none,
  1, 20, :_reduce_none,
  1, 20, :_reduce_none,
  3, 21, :_reduce_12,
  2, 21, :_reduce_13,
  1, 23, :_reduce_14,
  3, 23, :_reduce_15,
  1, 22, :_reduce_none,
  3, 22, :_reduce_17,
  3, 22, :_reduce_18 ]

racc_reduce_n = 19

racc_shift_n = 34

racc_token_table = {
  false => 0,
  :error => 1,
  ";" => 2,
  "," => 3,
  :OR_COMMA_OPERATOR => 4,
  :AND_OPERATOR => 5,
  :COMPARATOR => 6,
  :OPENING_BRACKET => 7,
  :CLOSING_BRACKET => 8,
  :UNRESERVED => 9,
  :SINGLE_QUOTE => 10,
  :SINGLE_QUOTED_STRING => 11,
  :DOUBLE_QUOTE => 12,
  :DOUBLE_QUOTED_STRING => 13 }

racc_nt_base = 14

racc_use_result_var = false

Racc_arg = [
  racc_action_table,
  racc_action_check,
  racc_action_default,
  racc_action_pointer,
  racc_goto_table,
  racc_goto_check,
  racc_goto_default,
  racc_goto_pointer,
  racc_nt_base,
  racc_reduce_table,
  racc_token_table,
  racc_shift_n,
  racc_reduce_n,
  racc_use_result_var ]

Racc_token_to_s_table = [
  "$end",
  "error",
  "\";\"",
  "\",\"",
  "OR_COMMA_OPERATOR",
  "AND_OPERATOR",
  "COMPARATOR",
  "OPENING_BRACKET",
  "CLOSING_BRACKET",
  "UNRESERVED",
  "SINGLE_QUOTE",
  "SINGLE_QUOTED_STRING",
  "DOUBLE_QUOTE",
  "DOUBLE_QUOTED_STRING",
  "$start",
  "target",
  "disjunction",
  "conjuction",
  "constraint",
  "selector",
  "argument",
  "array",
  "value",
  "contents" ]

Racc_debug_parser = false

##### State transition tables end #####

# reduce 0 omitted

# reduce 1 omitted

def _reduce_2(val, _values)
 create_logical_operator(:OR, val[0], val[2])
end

# reduce 3 omitted

def _reduce_4(val, _values)
 create_logical_operator(:AND, val[0], val[2])
end

# reduce 5 omitted

def _reduce_6(val, _values)
 create_constraint(val[0], val[1], val[2])
end

# reduce 7 omitted

def _reduce_8(val, _values)
 val[1]
end

# reduce 9 omitted

# reduce 10 omitted

# reduce 11 omitted

def _reduce_12(val, _values)
 val[1]
end

def _reduce_13(val, _values)
 []
end

def _reduce_14(val, _values)
 val
end

def _reduce_15(val, _values)
 val[0].push(val[2]); val[0]
end

# reduce 16 omitted

def _reduce_17(val, _values)
 val[1]
end

def _reduce_18(val, _values)
 val[1]
end

def _reduce_none(val, _values)
  val[0]
end

  end   # class Parser
end   # module RsqlRuby


